HDDS-9774. Store start time of datanode decommissioning

Tejaskriya opened the pull request:
## What changes were proposed in this pull request?
In order to keep a track of how long the decommissioning of a datanode has been running for, the time at which the datanode decommissioning was triggered is an important piece of information that must be available. At times it has been seen that this can't be found easily in the logs. Storing this data in the DatanodeadminMonitor will be helpful and convenient. This will later be displayed as a part of decommission-status command being introduced through [HDDS-9324](https://issues.apache.org/jira/browse/HDDS-9324)

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-9774

## How was this patch tested?

Existing unit tests and Added unit test under TestDatanodeAdminMonitor

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java:
```diff
@@ -78,7 +78,7 @@ public class DatanodeAdminMonitorImpl implements DatanodeAdminMonitor {
   private Queue<DatanodeDetails> pendingNodes = new ArrayDeque();
   private Queue<DatanodeDetails> cancelledNodes = new ArrayDeque();
-  private Set<DatanodeDetails> trackedNodes = new HashSet<>();
+  private Map<String, TrackedNode> trackedNodes = new HashMap<>();
```
Why does this need to be changed to a Map, rather than Set? The trackNode class has hashCode and equals based only on DatanodeDetails, so we could keep this as a Set. Maybe there is a reason you need to change it to a map I have missed?

Tejaskriya replied:
I have changed Set<DatanodeDetails> trackedNodes to Map<String, TrackedNode> trackednodes as in the future, we will want to fetch information of a particular datanode -> this would be difficult if we store it as a set as there is no get method (worst case, we'll need to iterate through the whole set and find a match).

Tejaskriya replied:
On second thoughts, keeping it as set will be better as most changes we are making to the tracked nodes are when we are iterating through the set. Moreover, even if we wish to fetch a particular DN, the set won't be very big so iterating through it to find one should be fine. I have reverted this back to set.

sodonnel commented:
> A disadvantage of this approach as it stands is that setting the time when we start tracking a node will likely be different from the time the decommission was actually triggered by a client. This is because the admin monitor is running in iterations separated by a period of time. So, suppose decommission was triggered at time `t`, then in the worst case the time may actually be set at `t + OZONE_SCM_DATANODE_ADMIN_MONITOR_INTERVAL`.

Could could resolve that by making pendingNode and cancelledNodes also use `TrackedNode` rather than just DatanodeDetails, then store the current time when the node is set into pending. It goes in there immediately, although it is not picked up for processing until the next iteration.

sodonnel commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java:
```diff
@@ -87,6 +87,44 @@ public class DatanodeAdminMonitorImpl implements DatanodeAdminMonitor {
+
+    private long startTime = 0L;
+
+    public TrackedNode(DatanodeDetails datanodeDetails) {
```
I might be a small improvement to make the constructor take both datanodeDetails and StartTime - then the object can be immutable and we can get rid of the `setStartTime()` method.

Tejaskriya commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java:
```diff
@@ -274,10 +323,10 @@ && checkContainersReplicatedOnNode(dn)) {
             } else if (status.isDecommissioning()) {
               completeDecommission(dn);
-              iterator.remove();
+              trackedNodes.remove(dn.getHostName());
```
I have changed this back to using an iterator now

Tejaskriya commented:
> A disadvantage of this approach as it stands is that setting the time when we start tracking a node will likely be different from the time the decommission was actually triggered by a client. 

Following the suggestion, I have changed pendingNodes To Queue<TrackedNodes> so when the startTime will no longer be noticeably different from the actual time decommissioning was triggered

@sodonnel @siddhantsangwan thank you for the first round of review! I have addressed all your comments. 
Please take a look at the new changes and leave any comments you have.

siddhantsangwan reviewed (commented):
@Tejaskriya can you please merge the latest master branch into this?

sodonnel reviewed (approved):
LGTM, pending green CI. I have kicked off the build now.

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java:
```diff
@@ -143,13 +181,32 @@ public synchronized void setMetrics(NodeDecommissionMetrics metrics) {
+   * Get a node which is currently tracked in the decommissioned
+   * or maintenance workflow.
+   *
+   * @return An unmodifiable set of the tracked nodes.
```
Javadoc for `@return` needs to be updated.

Tejaskriya commented:
> LGTM, pending green CI. I have kicked off the build now.

Thank you! There was a minor issue after merging the latest master branch to this branch. I have fixed it now. Could you please trigger the CI again?

siddhantsangwan commented on hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/node/DatanodeAdminMonitorImpl.java:
```diff
@@ -238,10 +295,11 @@ private void processPendingNodes() {
 
   private void processTransitioningNodes() {
```
The current code in `DatanodeAdminMonitorImpl` that reads and writes to the set `trackedNodes` is not thread safe. In this method, a thread is iterating through `trackedNodes` and can modify it by using the iterator. Meanwhile, another thread could read this set in the `getTrackedNodes` and `getSingleTrackedNode` methods. The latter two methods are synchronized, but `processTransitioningNodes` is not, which is incorrect. We could fix this by making this method synchronized. It's also possible to do fine grained synchronization by wrapping the `iterator.remove()` calls with synchronized. What do you think?

sodonnel replied:
What about just changing `trackedNodes` to be a concurrent Hash set `ConcurrentHashMap.newKeySet();`? That would avoid a potential problem of `processTransitioningNodes` taking a long time and lock out calls to the other methods while it is running.

siddhantsangwan replied:
Yeah, a concurrent set should also work well enough for our purpose I think.

Tejaskriya replied:
I have changed it to ConcurrentHashMap.newKeySet() in the recent commits. Could you check if it is fine now?

siddhantsangwan replied:
Looks fine. I've kicked off CI.

sodonnel commented:
All comments are addressed and CI is green so I am merging this. Thanks for the work and reviews.
